torch/csrc/jit/mobile/nnc/aot_compiler.cpp

```diff
@@ -189,6 +194,153 @@ std::pair<std::unique_ptr<Function>, const std::string> aotCompile(
 }
 
+void writeOutputLlvmAssembly(
+    const std::string& asm_code,
+    const std::string& output_llvm_file_name) {
+  std::ofstream output(output_llvm_file_name);
+  output << asm_code;
+  std::cout << "The compiled llvm assembly code was saved to "
+            << output_llvm_file_name << std::endl;
+}
+
+std::vector<std::string> split(
+    char separator,
+    const std::string& string,
+    bool ignore_empty = true) {
+  std::vector<std::string> pieces;
+  std::stringstream ss(string);
+  std::string item;
+  while (getline(ss, item, separator)) {
+    if (!ignore_empty || !item.empty()) {
+      pieces.push_back(std::move(item));
+    }
+  }
+  return pieces;
+}
+
+std::vector<std::vector<int64_t>> parseInputShapes(const std::string& input_dims_s) {
+  std::vector<std::string> input_dims_list = split(';', input_dims_s);
+  std::vector<std::vector<int64_t>> inputs;
+  for (const auto& input_dims_item : input_dims_list) {
+    auto input_dims_str = split(',', input_dims_item);
+    std::vector<int64_t> input_dims;
+    input_dims.reserve(input_dims_str.size());
+    for (const auto& s : input_dims_str) {
+      input_dims.push_back(c10::stoi(s));
+    }
+    inputs.push_back(input_dims);
+  }
+  return inputs;
+}
+
+std::vector<at::ScalarType> parseInputTypes(const std::string& input_types_str) {
+  std::vector<std::string> inputTypes = split(';', input_types_str);
+  std::vector<at::ScalarType> scalarTypes;
+  for (const auto& inputType : inputTypes) {
+    at::ScalarType scalarType;
+    if (inputType == "float") {
+      scalarType = at::ScalarType::Float;
+    } else if (inputType == "uint8") {
+      scalarType = at::ScalarType::Byte;
+    } else if (inputType == "int64") {
+      scalarType = at::ScalarType::Long;
+    } else {
+      CAFFE_THROW("Unsupported input type: ", inputType);
+    }
+    scalarTypes.push_back(scalarType);
+  }
+  return scalarTypes;
+}
+
+std::string getNncKernelId(
+    const std::string& model_name,
+    const std::string& model_version,
+    const std::string& method_name) {
+  // TODO: calculate the version_token.
+  const std::string version_token = "VERTOKEN";
+  return model_name + ":" + model_version + ":" + method_name + ":" +
+      version_token;
+}
+
+std::string getNncKernelFuncName(
+    const std::string& model_name,
+    const std::string& model_version,
+    const std::string& method_name) {
+  return "nnc_" + model_name + "_" + model_version + "_" + method_name;
+}
+
+std::shared_ptr<Graph> preprocessGraphPasses(
+    std::shared_ptr<Graph>& graph,
+    const std::vector<c10::optional<at::Tensor>>& example_inputs) {
+  torch::jit::RemoveTensorMutation(graph);
+  torch::jit::EliminateDeadCode(graph->block());
+  graph = torch::jit::tensorexpr::removeUnusedSelfArgument(graph);
+
+  torch::jit::tensorexpr::annotateInputShapes(graph, example_inputs);
+  torch::jit::OptimizeFrozenGraph(graph, true);
+  torch::jit::PropagateShapesOnGraph(graph);
```

Comment:
Do you think we can remove the shape propagation passes as TraceGraph() is called from aotCompile() ?

Reply:
I'd prefer to keep it (but I don't feel strongly about it). My motivation is that 1) it doesn't hurt to run shape-prop if all the shapes are already there, 2) in theory it's possible to invoke `preprocess` from some other place than `aotCompile` - in that case shape-prop can actually do work.